### PR TITLE
Add alerts card

### PR DESCRIPTION
**Pull request by pcbailey**
This PR adds the alerts card to the dashboard overview page. The alerts card does not display if alert-manager is not available.

Loaded alerts card:
Selection_567

Alerts card in loading state:
Selection_568

@rawagner

### CODE REVIEW

File: frontend/public/components/dashboard/alerts-card/alert-item.tsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  return (
+    <div className="co-alert__item">
+      {getSeverityIcon(getAlertSeverity(alert))}
+      <div className="co-alert__item-message">{getAlertDescription(alert) || getAlertMessage(alert)}</div>
+    </div>
+  );
+};
+
+type AlertItemProps = {
+  alert: object;
```
Comment (rawagner) [inline frontend/public/components/dashboard/alerts-card/alert-item.tsx]: this should have a better type. Can you reuse `Alert` from `monitoring.tsx` ?

File: frontend/public/components/dashboard/alerts-card/alerts-card.scss [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+.co-alert__icon--critical {
+  color: #c9190b;
+}
+
+.co-alert__icon--warning {
+  color: #f0ab00;
+}
+
+.co-alert__top-margin {
+  margin-top: -15px;
```
Comment (rawagner) [inline frontend/public/components/dashboard/alerts-card/alerts-card.scss]: This is not a good way to do this.
If `AlertsCard` is not shown, it will cause spacing issues (empty `GridItem` is still rendered).
I'd just add alerts into `HealthCard`, so we wont have `AlertsCard`, but all will be part of `Health`

File: frontend/public/components/dashboard/alerts-card/utils.tsx [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import * as _ from 'lodash-es';
+
+export const getAlertSeverity = alert => _.get(alert, 'labels.severity');
```
Comment (rawagner) [inline frontend/public/components/dashboard/alerts-card/utils.tsx]: types

File: frontend/public/components/dashboards-page/overview-dashboard/alerts-card.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  DashboardCardTitle,
+} from '../../dashboard/dashboard-card';
+import { AlertsBody, AlertItem, filterAlerts } from '../../dashboard/alerts-card';
+import { DashboardItemProps, withDashboardResources } from '../with-dashboard-resources';
+import { RootState } from '../../../redux';
+import { featureReducerName, flagPending } from '../../../reducers/features';
+import { FLAGS } from '../../../const';
+import { ALERTS_KEY } from '../../../actions/dashboards';
+
+export const ALERT_MANAGER_BASE_PATH = window.SERVER_FLAGS.alertManagerBaseURL;
```
Comment (rawagner) [inline frontend/public/components/dashboards-page/overview-dashboard/alerts-card.tsx]: why is this here ?

File: frontend/public/actions/dashboards.ts [hunk 5]
```diff
@@ -71,15 +73,27 @@ export const watchURL: WatchURLAction = (url, fetch = coFetchJSON) => (dispatch,
   }
 };
 
+export const watchAlerts: WatchAlertsAction = () => (dispatch, getState) => {
+  const isActive = isWatchActive(getState().dashboards, RESULTS_TYPE.ALERTS, ALERTS_KEY);
+  dispatch(activateWatch(RESULTS_TYPE.ALERTS, ALERTS_KEY));
+  if (!isActive) {
+    const alertMgrURL = `${window.SERVER_FLAGS.alertManagerBaseURL}/api/v2/alerts?silenced=false&inhibited=false`;
```
Comment (rawagner) [inline frontend/public/actions/dashboards.ts]: we dont even need to attempt fetching alerts if `window.SERVER_FLAGS.alertManagerBaseURL` is not defined. Take a look at `watchPrometheusQuery`

File: frontend/public/components/dashboard/alerts-card/alerts-card.scss [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+.co-alert__item-message {
+  display: table-cell;
+}
+
+.co-alert__icon {
+  margin-right: 1.5em;
+}
+
+.co-alert__icon--critical {
+  color: #c9190b;
```
Comment (rawagner) [inline frontend/public/components/dashboard/alerts-card/alerts-card.scss]: pf color variable

File: frontend/public/components/dashboard/alerts-card/alerts-card.scss [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+.co-alert__icon--warning {
+  color: #f0ab00;
```
Comment (rawagner) [inline frontend/public/components/dashboard/alerts-card/alerts-card.scss]: pf color variable

File: frontend/public/components/dashboard/alerts-card/alert-item.tsx (hunk 1 above)
Comment (pcbailey) [reply frontend/public/components/dashboard/alerts-card/alert-item.tsx]: The `Alert` type from `monitoring.tsx` worked perfectly. Thanks!

File: frontend/public/components/dashboard/alerts-card/alerts-card.scss (hunk 2 above)
Comment (pcbailey) [reply frontend/public/components/dashboard/alerts-card/alerts-card.scss]: Good point! Done!

File: frontend/public/components/dashboard/alerts-card/utils.tsx (hunk 3 above)
Comment (pcbailey) [reply frontend/public/components/dashboard/alerts-card/utils.tsx]: Done!

File: frontend/public/components/dashboards-page/overview-dashboard/alerts-card.tsx (hunk 4 above)
Comment (pcbailey) [reply frontend/public/components/dashboards-page/overview-dashboard/alerts-card.tsx]: This must have been left over from some stuff I had tried previously. It's been deleted now. Thanks for catching it!

File: frontend/public/actions/dashboards.ts (hunk 5 above)
Comment (pcbailey) [reply frontend/public/actions/dashboards.ts]: Done!

File: frontend/public/components/dashboard/alerts-card/alerts-card.scss (hunk 6 above)
Comment (pcbailey) [reply frontend/public/components/dashboard/alerts-card/alerts-card.scss]: Done!

File: frontend/public/components/dashboard/alerts-card/alerts-card.scss (hunk 7 above)
Comment (pcbailey) [reply frontend/public/components/dashboard/alerts-card/alerts-card.scss]: Done!

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx [hunk 8]
```diff
@@ -149,6 +165,21 @@ const HealthCard_: React.FC<HealthProps> = ({
...
+
+      {alerts.length > 0 &&
+        <React.Fragment>
+          <DashboardCardHeader className="co-alerts-card__border">
+            <DashboardCardTitle>Alerts</DashboardCardTitle>
+          </DashboardCardHeader>
+          <DashboardCardBody isLoading={flagPending(openShiftFlag)}>
+            <AlertsBody>
+              {alerts.map((alert, index) => (
+                <AlertItem key={`alert-${index}`} alert={alert} />
```
[the remark this answers is not in the document]
Comment (rawagner) [reply frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: `alert` has no id, lets go with just `index` then

File: frontend/public/components/dashboards-page/with-dashboard-resources.tsx [hunk 9]
```diff
@@ -109,10 +127,13 @@ type WithDashboardResourcesState = {
 type WithDashboardResourcesProps = {
   watchURL: WatchURLAction;
   watchPrometheusQuery: WatchPrometheusQueryAction;
+  watchAlerts: WatchAlertsAction;
   stopWatchURL: StopWatchURLAction;
   stopWatchPrometheusQuery: StopWatchPrometheusAction;
+  stopWatchAlerts: StopWatchAlertsAction;
   [RESULTS_TYPE.PROMETHEUS]: ImmutableMap<string, any>;
   [RESULTS_TYPE.URL]: ImmutableMap<string, any>;
+  [RESULTS_TYPE.ALERTS]: any;
```
Comment (rawagner) [inline frontend/public/components/dashboards-page/with-dashboard-resources.tsx]: should be `ImmutableMap<string, any>`

File: frontend/public/components/dashboards-page/with-dashboard-resources.tsx [hunk 10]
```diff
@@ -123,8 +144,11 @@ export type DashboardItemProps = {
   stopWatchURL: StopWatchURL;
   watchPrometheus: WatchPrometheus;
   stopWatchPrometheusQuery: StopWatchPrometheus;
+  watchAlerts: WatchAlerts;
+  stopWatchAlerts: StopWatchAlerts;
   urlResults: ImmutableMap<string, any>;
   prometheusResults: ImmutableMap<string, any>;
+  alertsResults: any;
```
Comment (rawagner) [inline frontend/public/components/dashboards-page/with-dashboard-resources.tsx]: same here `ImmutableMap<string, any>`

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (rawagner) [reply frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: maybe we could use `alert.alert` property. @pcbailey can you try ?

File: frontend/public/components/dashboard/health-card/health-card.scss [hunk 11]
```diff
@@ -45,3 +45,35 @@
 .co-health-card__title {
   padding-left: 0.6rem;
 }
+
```
[the remark this answers is not in the document]
Comment (pcbailey) [reply frontend/public/components/dashboard/health-card/health-card.scss]: I'm sorry. I wasn't aware of that. Fixed!

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (pcbailey) [reply frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: This has been changed to `alert.fingerprint` as discussed offline with @rawagner.

File: frontend/public/components/dashboards-page/with-dashboard-resources.tsx (hunk 9 above)
Comment (pcbailey) [reply frontend/public/components/dashboards-page/with-dashboard-resources.tsx]: Fixed!

File: frontend/public/components/dashboards-page/with-dashboard-resources.tsx (hunk 10 above)
Comment (pcbailey) [reply frontend/public/components/dashboards-page/with-dashboard-resources.tsx]: Fixed!

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx [hunk 12]
```diff
@@ -149,6 +165,21 @@ const HealthCard_: React.FC<HealthProps> = ({
...
+
+      {alerts.length > 0 &&
+        <React.Fragment>
+          <DashboardCardHeader className="co-health-card__alerts-border">
+            <DashboardCardTitle>Alerts</DashboardCardTitle>
+          </DashboardCardHeader>
+          <DashboardCardBody isLoading={flagPending(openShiftFlag)}>
+            <AlertsBody>
+              {alerts.map(alert => (
+                <AlertItem key={alert.fingerprint} alert={alert} />
```
Comment (cloudbehl) [inline frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: I think it should be `alert.annotations.fingerprint`. 

@rawagner thoughts?

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx [hunk 13]
```diff
@@ -57,6 +59,11 @@ const getK8sHealthState = (openShiftFlag: boolean, k8sHealth: any): SubsystemHea
     : { message: `${getName(openShiftFlag)} ${ERROR}`, state: HealthState.ERROR };
 };
 
+const getAlerts = (alertsResults: any): Alert[] => {
```
Comment (cloudbehl) [inline frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: can we move this to common place so other component leveraging can use this? @rawagner @pcbailey 
```suggestion
const getAlerts = (alertsResults: any): Alert[] => {
const alertsResponse = alertsResults ? alertsResults.getIn([ALERTS_KEY, 'result'], []) : [];
  return Array.isArray(alertsResponse) ? alertsResponse : [];
};
```

Comment (pcbailey) [reply frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: @cloudbehl Do you have any suggestions on a good location for it?

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx (hunk 12 above)
Comment (rawagner) [reply frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: Its correct AFAIK. `alert.fingerprint`

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx (hunk 13 above)
Comment (rawagner) [reply frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: i'd suggest `dashboard/health-card/utils.ts `

Comment (pcbailey) [reply frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: I thought about that, but I wasn't sure if that was "common" enough.

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx (hunk 12 above)
Comment (cloudbehl) [reply frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: @rawagner I verified its `alert.fingerprint`. Thanks

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx (hunk 13 above)
Comment (pcbailey) [reply frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: I have moved it to `dashboard/health-card/utils.ts`.

File: frontend/public/components/dashboard/health-card/health-card.scss [hunk 14]
```diff
@@ -1,3 +1,35 @@
+.co-health-card__alerts-body {
```
Comment (rawagner) [inline frontend/public/components/dashboard/health-card/health-card.scss]: please alphabetize these classes/rules

File: frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx [hunk 15]
```diff
@@ -9,7 +9,7 @@ import {
   DashboardCardTitle,
   DashboardCardSeeAll,
 } from '../../dashboard/dashboard-card';
-import { HealthBody, HealthItem } from '../../dashboard/health-card';
+import { HealthBody, HealthItem, AlertItem, AlertsBody, getAlerts } from '../../dashboard/health-card';
```
Comment (rawagner) [inline frontend/public/components/dashboards-page/overview-dashboard/health-card.tsx]: alpha

File: frontend/public/components/dashboards-page/with-dashboard-resources.tsx [hunk 16]
```diff
@@ -6,13 +6,18 @@ import { RESULTS_TYPE } from '../../reducers/dashboards';
 import {
   watchURL,
```
Comment (rawagner) [inline frontend/public/components/dashboards-page/with-dashboard-resources.tsx]: can you alphabetize these too ?
